src/main/java/com/artipie/rpm/ReactiveLock.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019 Yegor Bugayenko
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.rpm;
+
+import io.reactivex.rxjava3.core.Completable;
+import io.reactivex.rxjava3.core.CompletableEmitter;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Reactive sync primitive.
+ *
+ * @since 0.1
+ */
+class ReactiveLock {
+
+    /**
+     * Unlock and emit sync object.
+     */
+    private final Object sync = new Object();
+
+    /**
+     * Current state of the lock.
+     */
+    private final AtomicBoolean locked = new AtomicBoolean(false);
+
+    /**
+     * Queue of lock acquires.
+     */
+    private final Queue<CompletableEmitter> acquires = new ConcurrentLinkedQueue<>();
+
+    /**
+     * Acquire the lock in a reactive way.
+     *
+     * @return Completion when lock is acquired.
+     */
+    public Completable lock() {
+        return Completable.create(
+            emitter -> {
+                this.acquires.add(emitter);
+                this.tryToEmitNext();
+            }
+        );
+    }
+
+    /**
+     * Unlock the lock.
+     */
+    public void unlock() {
+        synchronized (this.sync) {
+            if (this.locked.compareAndSet(true, false)) {
+                this.tryToEmitNext();
+            }
+        }
+    }
+
+    /**
+     * Give a lock to another awaiter.
+     */
+    private void tryToEmitNext() {
+        synchronized (this.sync) {
```

Comment:
@Sammers21 this method can be rewritten using `compareAndSet` atomic operation of `AtomicBoolean`:
```java
if (this.locked.compareAndSet(false, true)) {
  final CompletableEmitter emitter = this.acquires.poll();
  if (emitter != null) {
       emitter.onComplete();
  }
}
```
so you can remove `sync` object, because `compareAndSet` is atomic operation.

Reply:
No, the approach is would cause deadlocks, by rewriting this way we would allow locking without emitting anything, so no one will perform the unlock operation.

Reply:
@Sammers21 then you can use:
```java
  if (emitter != null) {
       emitter.onComplete();
  } else {
    this.locked.set(false);
  }
```
I mean that combination of `synchronized (lock)` with `AtmoicBoolean` is useless. In other words you can use `private volatile boolean locked` instead of `AtomicBoolean`.

Reply:
Let's keep it for the purpose of readability.